frontends/PyCDE/src/pycde/value.py

```diff
@@ -258,6 +258,127 @@ def pad_or_truncate(self, num_bits: int):
+    will be truncated or sign-extended to that width.
+    """
+    return self._exec_cast(SignedBitVectorValue, ir.IntegerType.get_signed,
+                           width)
+
+  def asUInt(self, width: int = None):
+    """
+    Returns this value as an unsigned integer. If 'width' is provided, this value
+    will be truncated or zero-padded to that width.
+    """
+    return self._exec_cast(UnsignedBitVectorValue, ir.IntegerType.get_unsigned,
+                           width)
+
+  #  === Infix operators ===
+
+  # Signless operations. These will all return signless values - a user is
+  # expected to reapply signedness semantics if needed.
+  def __eq__(self, other):
+    from .dialects import comb
+    return comb.EqOp(self.asInt(), other.asInt())
+
+  def __ne__(self, other):
+    from .dialects import comb
+    return comb.NeOp(self.asInt(), other.asInt())
+
+  def __and__(self, other):
+    from .dialects import comb
+    return comb.AndOp(self.asInt(), other.asInt())
+
+  def __or__(self, other):
+    from .dialects import comb
+    return comb.OrOp(self.asInt(), other.asInt())
+
+  def __xor__(self, other):
+    from .dialects import comb
+    return comb.XorOp(self.asInt(), other.asInt())
+
+  def __invert__(self):
+    from .dialects import comb
+    from .pycde_types import types
+    return self.asInt() ^ types.int(self.type.width)(-1)
+
+  # Generalized function for executing sign-aware binary operations. Performs
+  # a check to ensure that the operands have signedness semantics, and then calls
+  # the provided operator.
+  def __exec_signedness_binop__(self, other, op, op_symbol):
+    signlessOperand = None
+    if type(self) is BitVectorValue:
+      signlessOperand = "LHS"
+    elif type(other) is BitVectorValue:
+      signlessOperand = "RHS"
+
+    if signlessOperand is not None:
+      raise TypeError(
+          f"Operator '{op_symbol}' is not supported on signless values. {signlessOperand} operand should be cast .asSInt()/.asUInt()."
+      )
+
+    return op(self, other)
+
+  def __add__(self, other):
+    from .dialects import hwarith
+    return self.__exec_signedness_binop__(other, hwarith.AddOp, "+")
+
+  def __sub__(self, other):
+    from .dialects import hwarith
+    return self.__exec_signedness_binop__(other, hwarith.SubOp, "-")
+
+  def __mul__(self, other):
+    from .dialects import hwarith
+    return self.__exec_signedness_binop__(other, hwarith.MulOp, "*")
+
+  def __truediv__(self, other):
+    from .dialects import hwarith
+    return self.__exec_signedness_binop__(other, hwarith.DivOp, "/")
+
+
+class WidthExtendingBitVectorValue(BitVectorValue):
+  # TODO: This class will contain comparison operators (<, >, <=, >=)
+  pass
+
+  def __lt__(self, other):
+    assert (False, "Unimplemented")
+
+  def __le__(self, other):
+    assert (False, "Unimplemented")
+
+  def __ge__(self, other):
+    assert (False, "Unimplemented")
+
+
+class UnsignedBitVectorValue(WidthExtendingBitVectorValue):
+  pass
```

Comment:
Does it make sense to implement a __neg__ and automatically convert so a signed int 1 bit wider?

Reply:
I wouldn't be a fan - I think the cases where a designer wants to use negation are already fairly limited, and negating unsigned values even more so. For those few cases, I think it is smarter to require the designer to be intentional about a value being signed if negation is needed.